Cilsil/Cil/Parsers/ExceptionParser.cs

```diff
@@ -0,0 +1,304 @@
+                (instruction, _) = state.PopInstruction();
+
+                /* Create exception allocation call and branch for handling exception. 
+                If exception allocation is True, store allocated exception expression as caught 
+                exception varaible. For example:
+
+                node2 preds:1 succs:4 exn: Conditional exception handler
+                n$49=_fun___instanceof(n$48:java.io.IOException*,sizeof(t=java.io.IOException):void)
+                PRUNE(n$49, true);
+                *&amp;CatchVar65:void=n$48
+
+                node3 preds:1 succs:5 exn: Conditional exception handler -- Connects to outter/finally block code directly
+                n$49=_fun___instanceof(n$48:java.io.IOException*,sizeof(t=java.io.IOException):void)
+                PRUNE(!n$49, false);*/
+
+                // Create exception allocation call:
+                // n${i+1}=_fun___instanceof(n${i}:{Exception_Type}*,sizeof(t={Exception_Type}:void)
+                (var exceptionAllocationCall, var exceptionObject) =
+                        CreateExceptionAllocationCall(state, Typ.FromTypeReference(exceptionType));
+
+                var trueNodeCondition = exceptionObject;
+                var falseNodeCondition = new UnopExpression(
+                    UnopExpression.UnopKind.LNot, trueNodeCondition, null);
+
+                // Create exception handling branch. True if exception is caught. False otherwise.
+                (var exceptionTrueNode, var exceptionFalseNode) =
+                    CreateExceptionBranchConditionNodes(state,
+                                                        trueNodeCondition,
+                                                        falseNodeCondition);
+
+                exceptionTrueNode.Instructions.Insert(0, exceptionAllocationCall);
+                exceptionFalseNode.Instructions.Insert(0, exceptionAllocationCall);
+
+                var catchVarIdentifier = state.GetIdentifier(Identifier.IdentKind.Normal);
+                var catchVarStore = new Store(catchVariable,
+                                              new VarExpression(returnVariable),
+                                              expressionType,
+                                              state.CurrentLocation);
+                exceptionTrueNode.Instructions.Add(catchVarStore);
+
+                RegisterNode(state, exceptionTrueNode);
+                RegisterNode(state, exceptionFalseNode);
+
+                Instruction nextInstructionIfNoException = null;
+
+                foreach (var singleInstruction in state.Method.Body.Instructions)
+                {
+                    if (singleInstruction.Offset >= endBlockOffset + 1)
+                    {
+                        nextInstructionIfNoException = singleInstruction;
+                        break;
+                    }
+                }
+                state.PushExpr(new ExnExpression(new VarExpression(returnVariable)), expressionType);
```

Comment:
Hmmm. are we sure this is generally the correct approach? I notice that EXN expression seems to generally be used  when the return/exit from the function context is an exception, in which case the expression represents the allocated exception and EXN is used to denote that the object is an exception.

Reply:
for example i noticed that in a separate context that "throw Exception()" is translated to a return of an EXN expression.

Reply:
EXN expression happens whenever there is exception handling outside of the method such as, for example, throw and the false branch in catch block like this:
```
n$6:System.Exception*=_fun static BuiltIn.__instanceof() (n$5:System.Exception*, sizeof(System.Exception, exact):void) [Line 89];
PRUNE (!(n$6):, False) [Line 89];
``` 
So we create EXN expression and push to stack in those cases. EXN expression got returned in catch block false branch because it is connected to the ret instruction directly.  I will make sure that in catch block the false branch connects to ret instruction only.

Reply:
Another interesting cases drew my attention are catch-and-non-empty-finally case and catch-and-empty-finally. In catch-and-non-empty-finally, since every catch has to go through finally, EXN expression for catch block false branch actually got returned **indirectly by connecting to finally block head node like the following node first**:
```
n$20=*&return:java.io.FileOutputStream* [line 65];
*&return:java.io.FileOutputStream*=null [line 65];
n$21=_fun___unwrap_exception(n$20:java.io.FileOutputStream*) [line 65];
```
For catch-and-empty-finally, it got even more tricky. Since finally block is empty and nothing really got processed, in this special corner case EXN expression for catch block false branch actually got returned **both directly by connecting to ret statement node and indirectly by connecting to finally block head node first**.